banyan/src/forest/mod.rs

```diff
@@ -5,33 +5,67 @@ use anyhow::Result;
 use core::{fmt::Debug, hash::Hash, iter::FromIterator, marker::PhantomData, ops::Range};
 use futures::future::BoxFuture;
 use libipld::cbor::DagCbor;
-use lru::LruCache;
+use parking_lot::Mutex;
 use rand::RngCore;
-use std::{
-    fmt::Display,
-    sync::{Arc, Mutex},
-};
+use std::{fmt::Display, num::NonZeroUsize, sync::Arc};
+use weight_cache::{Weighable, WeightCache};
 mod read;
 mod stream;
 mod write;
 pub(crate) use read::ForestIter;
 
 pub type FutureResult<'a, T> = BoxFuture<'a, Result<T>>;
 
+impl<T: TreeTypes> Weighable for Branch<T> {
+    fn measure(value: &Self) -> usize {
+        let mut bytes = std::mem::size_of::<Branch<T>>();
+        for child in value.children.iter() {
+            bytes += std::mem::size_of::<Index<T>>();
+            match child {
+                Index::Leaf(leaf) => {
+                    bytes += leaf.keys.estimated_size();
+                }
+                Index::Branch(branch) => {
+                    bytes += branch.summaries.estimated_size();
+                }
+            }
+        }
+        bytes
+    }
+}
+
+type CacheOrBypass<T> = Option<Arc<Mutex<WeightCache<<T as TreeTypes>::Link, Branch<T>>>>>;
+
 #[derive(Debug, Clone)]
-pub struct BranchCache<T: TreeTypes>(Arc<Mutex<lru::LruCache<T::Link, Branch<T>>>>);
+pub struct BranchCache<T: TreeTypes>(CacheOrBypass<T>);
+
+impl<T: TreeTypes> Default for BranchCache<T> {
+    fn default() -> Self {
+        Self::new(64 << 20)
+    }
+}
 
 impl<T: TreeTypes> BranchCache<T> {
+    /// Passing a capacity of 0 disables the cache.
     pub fn new(capacity: usize) -> Self {
-        Self(Arc::new(Mutex::new(LruCache::new(capacity))))
+        let cache = if capacity == 0 {
+            None
+        } else {
+            Some(Arc::new(Mutex::new(WeightCache::new(
+                NonZeroUsize::new(capacity).expect("Cache capacity must be "),
```

Comment:
you probably tried this and it looks reasonable, but reading the message alone in source code feels funny :-), quite existential …